compiler/rustc_resolve/src/imports.rs

```diff
@@ -1161,46 +1168,53 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                     msg,
                 );
             } else {
-                let error_msg = if crate_private_reexport {
-                    format!(
-                        "`{}` is only public within the crate, and cannot be re-exported outside",
-                        ident
-                    )
-                } else {
-                    format!("`{}` is private, and cannot be re-exported", ident)
-                };
-
                 if ns == TypeNS {
-                    let label_msg = if crate_private_reexport {
-                        format!("re-export of crate public `{}`", ident)
+                    let mut err = if crate_private_reexport {
+                        let mut err = self.tcx.sess.create_err(CannotBeReexportedCratePublicNS {
+                            span: import.span,
+                            ident,
+                        });
+                        err.subdiagnostic(ReexportOfCratePublic {
+                            span: import.span,
+                            ident: ident,
+                        });
+                        err
                     } else {
-                        format!("re-export of private `{}`", ident)
+                        let mut err = self
+                            .tcx
+                            .sess
+                            .create_err(CannotBeReexportedPrivateNS { span: import.span, ident });
+                        err.subdiagnostic(ReexportOfPrivate { span: import.span, ident: ident });
```

Comment:
Going to sleep right now. Sorry for the nits, but I don't think this needs to use `subdiagnostic`. You could just attach `#[help(resolve_consider_adding_macro_export)]` on the main struct since they use the same fields. I don't mind landing this and then I can make a PR that addresses this, so that you don't have to worry about this PR. because CI is still running I can't approve this rn.

Reply:
No worries, nits are appreciated to help my understanding of what's expected :>

I'll push this in a moment